8159055: ImageIcon.setImage can't handle null parameter

prsadhuk opened the pull request:
When trying to call 'icon.setImage(null);' where 'icon' is an instance of ImageIcon, a null pointer exception is thrown at runtime.
The code tried to get the `id` for that image and instantiates `MediaTracker` to associate the null image to that `id` and checks the status of loading this null image, removes the null image from the tracker and then tries to get the image width where it throws NPE as image is null.

It's better to not go through all MediaTracker usage and bail out initially itself for null image..

kumarabhi006 commented:
Does it require to update the doc as well to explicitly mention about NULL parameter ?
Otherwise fix looks ok to me.

prsadhuk commented:
I feel it's better to inform the user in the spec otherwise application may not know what will be done with NULL image

prsadhuk commented:
yes..i will raise it once we are done with this review approval..

kumarabhi006 reviewed (approved):
LGTM

prsadhuk commented:
Raised CSR https://bugs.openjdk.org/browse/JDK-8359398
@azuev-java @kumarabhi006 Please review.

prsadhuk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -367,9 +367,13 @@ public Image getImage() {
      * @param image the image
      */
     public void setImage(Image image) {
+        if (image == null) {
```
In constructor also, it returns

https://github.com/openjdk/jdk/blob/991097b7bf08cc1a4ceedb0c555b12948ae71885/src/java.desktop/share/classes/javax/swing/ImageIcon.java#L146-L148
https://github.com/openjdk/jdk/blob/991097b7bf08cc1a4ceedb0c555b12948ae71885/src/java.desktop/share/classes/javax/swing/ImageIcon.java#L185-L187

prsadhuk replied:
ok

prsadhuk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -367,10 +367,14 @@ public Image getImage() {
 
     /**
      * Sets the image displayed by this icon.
+     * A null image is ignored.
      * @param image the image
```
I dont think so..Till now, it was throwing NPE so it was not expected so it cannot be "may be null", which is we are telling the application to avoid..

aivanov-jdk commented on test/jdk/javax/swing/ImageIcon/ImageIconNullImageTest.java:
```diff
@@ -0,0 +1,42 @@
+public class ImageIconNullImageTest {
+
+    public static void main(String[] args) throws Exception {
+        SwingUtilities.invokeAndWait(() -> {
+            ImageIcon icon = new ImageIcon();
+            icon.setImage(null);
+        });
```
`invokeAndWait` isn't needed here, you can run the test code on the main thread.

A comment that no NPE is expected would be good to clarify what the test expects.

aivanov-jdk reviewed (changes requested):
The `ImageIcon` constructor that accepts `Image` has the same problem, or a similar problem.

https://github.com/openjdk/jdk/blob/9d060574e5dbd13e634f00d749d0108ceff1fae8/src/java.desktop/share/classes/javax/swing/ImageIcon.java#L224-L226

It throws `NullPointerException` if a `null` parameter is passed. Other constructors handle `null` values, so you should unify handling of `null` images.

aivanov-jdk commented on test/jdk/javax/swing/ImageIcon/ImageIconNullImageTest.java:
```diff
@@ -29,14 +29,13 @@
+
+        // Setting null image shouldn't cause NPE
+        icon.setImage(null);
     }
```
Add the test for `new ImageIcon((Image) null)`: currently it throws `NullPointerException` but it won't throw it after your fix.

prsadhuk replied:
ok

prsadhuk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -223,6 +223,9 @@ public ImageIcon(Image image, String description) {
      */
     public ImageIcon (Image image) {
```
I have expanded for those methods where there is code change for NULL interception....Not sure about others..

mrserb replied:
Note that "ImageIcon(URL location)" will throw npe as well, do we want to ignore it or specify NPE? That is actually interesting question since initially some of methods throw NPE and another ignore the image.

prsadhuk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -367,10 +370,14 @@ public Image getImage() {
+     * Setting null image will not render any image icon.
      * @param image the image
      */
     public void setImage(Image image) {
         this.image = image;
+        if (image == null) {
+            return;
+        }
```
Update the description from where? Its not overridden method..

aivanov-jdk replied:
Set it to `null` if the image is `null` or load it from the `"comment"` property as the constructor does.

However, it's safer to leave it untouched. After all, constructors are used to initialise the fields of the object so it makes sense to try to set a description whereas changing the icon doesn't necessarily mean the description needs to change; if it does, call `setDescription` explicitly.

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -367,10 +370,14 @@ public Image getImage() {
 
     /**
      * Sets the image displayed by this icon.
+     * Setting null image will not render any image icon.
```
```suggestion
     * Setting a {@code null} image will not render any image icon.
```

Will it sound clearer if the text stated, _“Setting a {@code null} image will remove the icon?”_

aivanov-jdk reviewed (approved):
Looks good to me.

I still have doubts about the description of how `null` is handled…

kumarabhi006 commented on test/jdk/javax/swing/ImageIcon/ImageIconNullImageTest.java:
```diff
@@ -0,0 +1,52 @@
+/*
+ * @test
+ * @bug 8159055
+ * @summary Verifies ImageIcon.setImage handles null parameter
```
Does it need to modify the summary as well ?

Apart from setImage method, null parameter is handled in constructor also.

kumarabhi006 commented:
I think JBS title can be modified to not to be specific to `setImage` method only and then CSR should be updated as well.

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -367,10 +371,14 @@ public Image getImage() {
 
     /**
      * Sets the image displayed by this icon.
+     * Setting a {@code null} image will not render any image icon.
```
```suggestion
     * Setting a {@code null} image renders no icon.
```
I think this is better: it's shorter, it uses the present tense because it's fact and it doesn't depend on anything.

prrace replied:
I think the existing text was OK. And "will" is better because it hasn't happened yet.
Rendering will happen later. Not when you construct this object.
My personally preferred wording would be something like
Setting a {@code null} image means any existing image will be removed and no image will be rendered.

Now .. what should happen to an existing description when you set a new image ?

Apps can call setDescription() but in what order ? 

The only thing I can say here is that I think if the image is null we probably should clear the description.

prsadhuk replied:
I have modified the wording as per suggestion and also clearing the description now if image is null..

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -215,6 +215,7 @@ public ImageIcon(Image image, String description) {
 
     /**
      * Creates an ImageIcon from an image object.
+     * Setting a {@code null} image will not render any image icon.
```
```suggestion
     * Passing a {@code null} image renders no icon.
```

aivanov-jdk commented on test/jdk/javax/swing/ImageIcon/ImageIconNullImageTest.java:
```diff
@@ -0,0 +1,52 @@
+/*
+ * @test
+ * @bug 8159055
+ * @summary Verifies ImageIcon setImage and constructor handles null parameter
```
```suggestion
 * @summary Verifies ImageIcon setImage and ImageIcon(Image) constructor handle null parameter
```
Let's be more specific.

prrace commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -223,6 +224,9 @@ public ImageIcon(Image image, String description) {
+        }
         Object o = image.getProperty("comment", imageObserver);
         if (o instanceof String) {
             description = (String) o;
```
Hmm. This is wasted work if the app calls ImageIcon(Image, String) because that promptly over-writes whatever was obtained via this code.

Also that other constructor (which github won't let me comment on) installs the description even if the image is null.
That seems to be inconsistent with everything else that has a null description for a null image.

aivanov-jdk replied:
> Also that other constructor (which github won't let me comment on) installs the description even if the image is null.

I don't see it as a problem.

If the app developer wants to initialise the object with `null` image and a description, why don't we let them do it?

The image may be auto-generated and while it's generated, the image in `ImageIcon` remains `null`; when the image is read the app calls `ImageIcon.setImage` to the set now available image and *expects the description being preserved*.

aivanov-jdk replied:
> Now .. what should happen to an existing description when you set a new image ?
>
> Apps can call setDescription() but in what order ?
>
> The only thing I can say here is that I think if the image is null we probably should clear the description.

Nothing should happen to the description. These are two fields which can be changed independently now, why would we enforce an order on method calls?

Changing the image shouldn't change the description; similarly, changing description doesn't change the image.

prrace replied:
> > Also that other constructor (which github won't let me comment on) installs the description even if the image is null.
> 
> I don't see it as a problem.
> 
> If the app developer wants to initialise the object with `null` image and a description, why don't we let them do it?
> 

Because that  would be inconsistent with every other constructor.

See
ImagIcon(String filename)
ImageIcon(String filename, String description)
ImageIcon(String URL)
ImageIcon(String URL, String description)
ImageIcon(byte[] imageData)
ImageIcon(byte[] imageData, String description)

None of these set the description if the image resolves to null.

Also clearly
ImageIcon(Image image) where image == null can't be doing it today
because a NPE is thrown before it gets there.
So we can have that case do whatever we want if we allow a null image.
And consistency seems the best choice.

prsadhuk replied:
If we leave it to throw NPE, then JCK might crib, if at some point they decide to add test to test null image and since javadoc doesn't mention it can throw NPE, it will result in non-compliance and then we have to modify javadoc to clarify it, so I guess it's best to take preventive action now..

aivanov-jdk replied:
> If we leave it to throw NPE, then JCK might crib, if at some point they decide to add test to test null image and since javadoc doesn't mention it can throw NPE, it will result in non-compliance and then we have to modify javadoc to clarify it, so I guess it's best to take preventive action now..

But it is *consistent* with the current behaviour and it is not specified at the moment.

If we explicitly specify the constructors throw NPE if a `null` parameter is passed, then JCK will add a test that confirms the specified behaviour. All the constructors come into picture then, which seems out of scope for this issue…

If we don't change the behaviour of constructors and JCK challenges the implementation, then such a JCK test has no foundation in the API specification; yet it also leads to JCK challenge of the clarity of the specification.

ExE-Boss reviewed (commented):
These are not needed as the initial value of `description` is `null`:

prsadhuk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -144,6 +144,7 @@ private static final Component createComponent() {
     public ImageIcon(String filename, String description) {
         image = Toolkit.getDefaultToolkit().getImage(filename);
         if (image == null) {
+            this.description = null;
```
We dont know the exact order of setDescription call..If that is called before setImage or constructor (unlikely but can be called), the description would not be null, so it's better to clear the description for null image to be safe..

aivanov-jdk replied:
> We dont know the exact order of setDescription call..

What do you mean?

It's a constructor, nothing else has been called on the instance yet.

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -367,10 +378,17 @@ public Image getImage() {
         this.image = image;
+        if (image == null) {
+            this.description = null;
+            return;
+        }
```
We must not change the description, but we should reset `width` and `height` to `-1`.
```suggestion
            width = -1;
            height = -1;
            return;
        }
```

aivanov-jdk commented:
/reviewers 2 reviewer

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -144,6 +144,7 @@ private static final Component createComponent() {
     public ImageIcon(String filename, String description) {
         image = Toolkit.getDefaultToolkit().getImage(filename);
         if (image == null) {
+            this.description = null;
```
And this code is in constructor, therefore `description` still has its default value of `null`.

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -367,10 +378,17 @@ public Image getImage() {
     public void setImage(Image image) {
         this.image = image;
+        if (image == null) {
+            this.description = null;
```
**I *strongly* disagree to changing the value of `description`** here, even though [I brought up this concern](https://github.com/openjdk/jdk/pull/25767#discussion_r2150249349).

This does not make sense. Consider the following code snippet:

```java
        ImageIcon im = new ImageIcon();
        // Both image and description are null

        im.setDescription("whatever");
        // Now image is null, and description isn't null

        im.setImage(null);
        // Why does description change to null?
```

Both `image` and `description` are two independent fields of `ImageIcon` object, each has its own getter and setter, and each can be changed independently. We should not enforce the order of calls: if an app developer wants to set the image to `null`, temporarily or not, the value of the description has to be preserved.

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -223,6 +225,9 @@ public ImageIcon(Image image, String description) {
+        if (image == null) {
+            return;
+        }
         Object o = image.getProperty("comment", imageObserver);
         if (o instanceof String) {
             description = (String) o;
```
To address [Phil's comment](https://github.com/openjdk/jdk/pull/25767/files#r2155610976):

> This is wasted work if the app calls ImageIcon(Image, String) because that promptly over-writes whatever was obtained via this code.

If we're going to change the constructors, to avoid *this wasted work* when `ImageIcon(Image, String)` constructor is called, I suggest moving the work into `ImageIcon(Image, String)` and implement `ImageIcon` like this:

```java
    public ImageIcon (Image image) {
        String description = null;
        if (image != null) {
            Object o = image.getProperty("comment", null);
            if (o instanceof String) {
                description = (String) o;
            }
        }
        this(image, description);
```
It is allowed in JDK 22 and later.

ExE-Boss replied:
Note that this was a preview language feature until **JDK 25**.

aivanov-jdk replied:
> Note that this was a preview language feature until **JDK 25**.

Huh… We're developing JDK 26 now, so the **Flexible Constructor Bodies** [[JEP 513](https://openjdk.org/jeps/513)] feature can safely be used.

ExE-Boss replied:
Yes, but it complicates backports.

aivanov-jdk replied:
> Yes, but it complicates backports.

Yes, it complicates backports, but this implies backports are possible. If the specification is updated, the backports will likely be impossible.

prsadhuk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -215,6 +215,8 @@ public ImageIcon(Image image, String description) {
     /**
      * Creates an ImageIcon from an image object.
+     * Setting a {@code null} image means
+     * no image will be rendered.
```
ok

prsadhuk commented:
> Both image and description are two independent fields of ImageIcon object, each has its own getter and setter, and each can be changed independently. We should not enforce the order of calls: if an app developer wants to set the image to null, temporarily or not, the value of the description has to be preserved

I guess resetting the description came from the fact that description is mentioned as "Sets the description of the image. This is meant to be a brief textual description of the object." so if image is null, the description of the image logically should be nothing.

But again, as pointed out, description has its own setter and getter so user would expect whatever description is stored either by [ImageIcon](https://docs.oracle.com/en/java/javase/24/docs/api/java.desktop/javax/swing/ImageIcon.html#%3Cinit%3E(java.awt.Image,java.lang.String))([Image](https://docs.oracle.com/en/java/javase/24/docs/api/java.desktop/java/awt/Image.html) image, [String](https://docs.oracle.com/en/java/javase/24/docs/api/java.base/java/lang/String.html) description) constructor or `setDescription `would be returned same via `getDescription`..
If not, the spec for set and getDescription needs to be modified to to incorporate null-image scenario.
But all-in-all I guess this description change can go in separate PR along with [above change](https://github.com/openjdk/jdk/pull/25767#discussion_r2164102520)

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -223,6 +225,9 @@ public ImageIcon(Image image, String description) {
      */
     public ImageIcon (Image image) {
         this.image = image;
+        if (image == null) {
+            return;
+        }
```
Are we changing the constructor to accept `null` as the `Image` parameter?

There's [a long thread](https://github.com/openjdk/jdk/pull/25767/files#r2155610976) about *consistent* behaviour of the `ImageIcon` constructors, yet no decision has been taken.

@prrace mentioned consistency among constructors, and `ImageIcon` constructors consistently throw `NullPointerException` for `null` image. In fact, he referred to setting the `description` field, but this is moot if a constructor throws.

aivanov-jdk commented:
> > Both image and description are two independent fields of ImageIcon object, each has its own getter and setter, and each can be changed independently. We should not enforce the order of calls: if an app developer wants to set the image to null, temporarily or not, the value of the description has to be preserved
> 
> I guess resetting the description came from the fact that description is mentioned as "Sets the description of the image. This is meant to be a brief textual description of the object." so if image is null, the description of the image logically should be nothing.

I agree that a description for a `null` image doesn't make sense.

Then, it also depends on how it's really handled. If the image in the `ImageIcon` object is `null`, the description could be ignored.

> But again, as pointed out, description has its own setter and getter so user would expect whatever description is stored either by [ImageIcon](https://docs.oracle.com/en/java/javase/24/docs/api/java.desktop/javax/swing/ImageIcon.html#%3Cinit%3E(java.awt.Image,java.lang.String))([Image](https://docs.oracle.com/en/java/javase/24/docs/api/java.desktop/java/awt/Image.html) image, [String](https://docs.oracle.com/en/java/javase/24/docs/api/java.base/java/lang/String.html) description) constructor or `setDescription `would be returned same via `getDescription`.. If not, the spec for set and getDescription needs to be modified to to incorporate null-image scenario.

Exactly.

Updating the specification for `setDescription` and `getDescription` to indicate the description will be set to `null` if the image is set to `null` will make the API cumbersome and less clear.

I prefer leaving the description intact.

> But all-in-all I guess this description change can go in separate PR along with [above change](https://github.com/openjdk/jdk/pull/25767#discussion_r2164102520)

Have anyone submit a bug to address this deficiency? If not, I'll submit one.

prrace commented:
I think we need to have the "whole picture" tested to make sure everything does as we expect.
It isn't so easy to tell even from looking at the source of ImageIcon.
Then spec'ed so developers know what to expect. 
When I wrote a little test, I see null args do generate NPEs but invalid args don't result in null images
Toolkit images are installed, even if they can't be used because the source isn't valid.
They are "effectively" null, but not "actually" null.
We should at least specify the null arg behaviours and try to say something about  invalid image data.
And as a result, I am now flipping to just documenting the setImage(null) NPE
The protected method loadImage probably needs to say it too.
Invalid image data, maybe we can cover that in a class level statement.
Note: I'd like to see every constructor tested with both null and invalid image data.

prsadhuk commented:
> I think we need to have the "whole picture" tested to make sure everything does as we expect. It isn't so easy to tell even from looking at the source of ImageIcon. Then spec'ed so developers know what to expect. When I wrote a little test, I see null args do generate NPEs but invalid args don't result in null images Toolkit images are installed, even if they can't be used because the source isn't valid. They are "effectively" null, but not "actually" null. We should at least specify the null arg behaviours and try to say something about invalid image data. And as a result, I am now flipping to just documenting the setImage(null) NPE The protected method loadImage probably needs to say it too. Invalid image data, maybe we can cover that in a class level statement. Note: I'd like to see every constructor tested with both null and invalid image data.

I tested with null image/imagedata and invalid imagedata and it seems only these constructors throw NPE (current JDK) for null image (invalid image doesnt throw any exception)

ImageIcon(image, desc)
ImageIcon(image)
ImageIcon(byte[], desc)
ImageIcon(byte[])

Current PR checks for null in ImageIcon(image) so it will not throw NPE but ImageIcon(byte[]) will still throw..
Should we handle those in separate PR and let this only be for setImage(null)?

aivanov-jdk commented:
> > I think we need to have the "whole picture" tested to make sure everything does as we expect. …\
> > When I wrote a little test, I see null args do generate NPEs but invalid args don't result in null images\
> > Toolkit images are installed, even if they can't be used because the source isn't valid. They are "effectively" null, but not "actually" null.
> 
> I tested with null image/imagedata and invalid imagedata and it seems only these constructors throw NPE (current JDK) for null image (invalid image doesnt throw any exception)

I wrote a JUnit test, it's more convenient than a jtreg test:

<details>
<summary><code>ImageIconConstructorsTest.java</code></summary>

```java
import java.awt.Image;
import java.net.URL;

import javax.swing.ImageIcon;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;

public class ImageIconConstructorsTest {
    @Test
    public void noArgs() {
        ImageIcon im = new ImageIcon();
        assertNull(im.getImage());
        assertNull(im.getDescription());
    }

    @Test
    public void stringNull() {
        ImageIcon im = new ImageIcon((String) null);
        assertNotNull(im.getImage());
        assertNull(im.getDescription());
    }

    @Test
    public void stringNullNull() {
        ImageIcon im = new ImageIcon((String) null, null);
        assertNotNull(im.getImage());
        assertNull(im.getDescription());
    }

    @Test
    public void stringNullWithDescription() {
        ImageIcon im = new ImageIcon((String) null, "Description");
        assertNotNull(im.getImage());
        assertEquals("Description", im.getDescription());
    }

    @Test
    public void byteArrayNull() {
        assertThrows(NullPointerException.class,
                     () -> new ImageIcon((byte[]) null));
//        ImageIcon im = new ImageIcon((byte[]) null);
//        assertNull(im.getImage());
//        assertNull(im.getDescription());
    }

    @Test
    public void byteArrayNullNull() {
        assertThrows(NullPointerException.class,
                     () -> new ImageIcon((byte[]) null, null));
//        ImageIcon im = new ImageIcon((byte[]) null, null);
//        assertNull(im.getImage());
//        assertNull(im.getDescription());
    }

    @Test
    public void byteArrayNullWithDescription() {
        assertThrows(NullPointerException.class,
                     () -> new ImageIcon((byte[]) null, "Description"));
//        ImageIcon im = new ImageIcon((byte[]) null, "Description");
//        assertNull(im.getImage());
//        assertNull(im.getDescription());
    }

    @Test
    public void urlNull() {
        assertThrows(NullPointerException.class,
                     () -> new ImageIcon((URL) null));
//        ImageIcon im = new ImageIcon((URL) null);
//        assertNull(im.getImage());
//        assertNull(im.getDescription());
    }

    @Test
    public void urlNullNull() {
        assertThrows(NullPointerException.class,
                     () -> new ImageIcon((URL) null, null));
//        ImageIcon im = new ImageIcon((URL) null, null);
//        assertNull(im.getImage());
//        assertNull(im.getDescription());
    }

    @Test
    public void urlNullWithDescription() {
        assertThrows(NullPointerException.class,
                     () -> new ImageIcon((URL) null, "Description"));
//        ImageIcon im = new ImageIcon((URL) null, "Description");
//        assertNull(im.getImage());
//        assertNull(im.getDescription());
    }

    @Test
    public void imageNull() {
        assertThrows(NullPointerException.class,
                     () -> new ImageIcon((Image) null));
//        ImageIcon im = new ImageIcon((Image) null);
//        assertNull(im.getImage());
//        assertNull(im.getDescription());
    }

    @Test
    public void imageNullNull() {
        assertThrows(NullPointerException.class,
                     () -> new ImageIcon((Image) null, null));
//        ImageIcon im = new ImageIcon((Image) null, null);
//        assertNull(im.getImage());
//        assertNull(im.getDescription());
    }

    @Test
    public void imageNullWithDescription() {
        assertThrows(NullPointerException.class,
                     () -> new ImageIcon((Image) null, "Description"));
//        ImageIcon im = new ImageIcon((Image) null, "Description");
//        assertNull(im.getImage());
//        assertNull(im.getDescription());
    }

}
```

</details>

All the test cases currently pass, and most constructors throw NPE for `null`. I haven't tested invalid arguments.

aivanov-jdk commented:
> > But all-in-all I guess this description change can go in separate PR along with [above change](https://github.com/openjdk/jdk/pull/25767#discussion_r2164102520)
> 
> Have anyone submit a bug to address this deficiency? If not, I'll submit one.

I submitted [JDK-8361610](https://bugs.openjdk.org/browse/JDK-8361610): _Avoid wasted work in ImageIcon(Image) for setting description_ for updating `ImageIcon(Image)` and `ImageIcon(Image, String)` constructors.

prrace commented:
> > I think we need to have the "whole picture" tested to make sure everything does as we expect. It isn't so easy to tell even from looking at the source of ImageIcon. Then spec'ed so developers know what to expect. When I wrote a little test, I see null args do generate NPEs but invalid args don't result in null images Toolkit images are installed, even if they can't be used because the source isn't valid. They are "effectively" null, but not "actually" null. We should at least specify the null arg behaviours and try to say something about invalid image data. And as a result, I am now flipping to just documenting the setImage(null) NPE The protected method loadImage probably needs to say it too. Invalid image data, maybe we can cover that in a class level statement. Note: I'd like to see every constructor tested with both null and invalid image data.
> 
> I tested with null image/imagedata and invalid imagedata and it seems only these constructors throw NPE (current JDK) for null image (invalid image doesnt throw any exception)
> 
> ImageIcon(image, desc) ImageIcon(image) ImageIcon(byte[], desc) ImageIcon(byte[])
> 
> Current PR checks for null in ImageIcon(image) so it will not throw NPE but ImageIcon(byte[]) will still throw.. Should we handle those in separate PR and let this only be for setImage(null)?

You told me (off-line) that null URL also throws NPE and I see that too.

This class is a bit of a mess of accidental behaviours and scant documentation.

Let's document all the NPE behaviours and include a test that verifies it.

And invalid data like "new byte[0]" or null for a file name,  or pointing to something that isn't an image file, or an invalid URL .. etc .. results in an Image instance being present, but it will never be completed so can never be drawn.

I think we need a class level statement like
"If the image source parameter to a constructor is non-null, but does not reference valid accessible image data, no exceptions will be thrown but the image will be 'effectively' null, as it will have no dimensions and never be drawn, and
getImageLoadStatus() will report ERRORED" - you should verify that last part but I think it is right.

We probably should add the similar text on setImage().

We can also test the invalid image data cases too.

prsadhuk commented:
> > > I think we need to have the "whole picture" tested to make sure everything does as we expect. It isn't so easy to tell even from looking at the source of ImageIcon. Then spec'ed so developers know what to expect. When I wrote a little test, I see null args do generate NPEs but invalid args don't result in null images Toolkit images are installed, even if they can't be used because the source isn't valid. They are "effectively" null, but not "actually" null. We should at least specify the null arg behaviours and try to say something about invalid image data. And as a result, I am now flipping to just documenting the setImage(null) NPE The protected method loadImage probably needs to say it too. Invalid image data, maybe we can cover that in a class level statement. Note: I'd like to see every constructor tested with both null and invalid image data.
> > 
> > 
> > I tested with null image/imagedata and invalid imagedata and it seems only these constructors throw NPE (current JDK) for null image (invalid image doesnt throw any exception)
> > ImageIcon(image, desc) ImageIcon(image) ImageIcon(byte[], desc) ImageIcon(byte[])
> > Current PR checks for null in ImageIcon(image) so it will not throw NPE but ImageIcon(byte[]) will still throw.. Should we handle those in separate PR and let this only be for setImage(null)?
> 
> You told me (off-line) that null URL also throws NPE and I see that too.
> 
> This class is a bit of a mess of accidental behaviours and scant documentation.
> 
> Let's document all the NPE behaviours and include a test that verifies it.
> 
> And invalid data like "new byte[0]" or null for a file name, or pointing to something that isn't an image file, or an invalid URL .. etc .. results in an Image instance being present, but it will never be completed so can never be drawn.
> 
> I think we need a class level statement like "If the image source parameter to a constructor is non-null, but does not reference valid accessible image data, no exceptions will be thrown but the image will be 'effectively' null, as it will have no dimensions and never be drawn, and getImageLoadStatus() will report ERRORED" - you should verify that last part but I think it is right.
> 
> We probably should add the similar text on setImage().
> 
> We can also test the invalid image data cases too.

Documented all NPE behavour and added class-level statement.
Updated test to verify it.
getImageLoadStatus() does return 4 which is ERRORED for invalid image, updated test reflects that.

prrace commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -367,10 +378,18 @@ public Image getImage() {
      */
     public void setImage(Image image) {
         this.image = image;
+        if (image == null) {
```
Umm. We had decided (in the end) to document the NPE.

prsadhuk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -217,6 +225,7 @@ public ImageIcon(Image image, String description) {
      * Creates an ImageIcon from an image object.
      * If the image has a "comment" property that is a string,
      * then the string is used as the description of this icon.
+     * Passing {@code null} Image will result in {@code NullPointerException}.
```
fixed

prsadhuk commented:
Test updated

prrace commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -61,6 +61,11 @@
  * from Images. Images that are created from a URL, filename or byte array
  * are preloaded using MediaTracker to monitor the loaded state
  * of the image.
+ * If the image source parameter to a constructor is non-null,
```
Whilst reviewing the CSR I realized you need to change the line above to say
"If the image source parameter to a constructor or method is non-null,"

I made that change in the CSR. Do it here too. It is needed to cover setImage()

prsadhuk replied:
Added

prsadhuk commented on test/jdk/javax/swing/ImageIcon/ImageIconTest.java:
```diff
@@ -0,0 +1,128 @@
+                    if (expected) {
+                        passed = true;
+                    }
+                } catch (Exception ex) {
```
I added this to handle if "invalid data" handling causes the code to throw IOException or something like that..
But anyways removed now..

prsadhuk commented on test/jdk/javax/swing/ImageIcon/ImageIconTest.java:
```diff
@@ -0,0 +1,128 @@
+        File file = new File(System.getProperty("test.src", "."), imgName);
+
+        for (ArgType a : ArgType.values()) {
+            for (final boolean invalid : new boolean[]{false, true}) {
```
Updated

ExE-Boss commented on test/jdk/javax/swing/ImageIcon/ImageIconTest.java:
```diff
@@ -39,63 +39,56 @@
 public class ImageIconTest {
 
     static enum ArgType { FILE, URL, BYTEARRAY, IMAGE };
+    static enum ArgVal { NULL, INVALIDDATA };
```
Maybe name this `INVALID_DATA`:
```suggestion
    static enum ArgVal { NULL, INVALID_DATA };
```

prsadhuk replied:
ok

prrace commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -176,6 +181,7 @@ public ImageIcon (String filename) {
      * Creates an ImageIcon from the specified URL. The image will
      * be preloaded by using MediaTracker to monitor the loaded state
      * of the image.
+     * Passing {@code null} URL will result in {@code NullPointerException}.
```
In the CSR, Joe quite reasonably requested that these all become @throws clauses

prrace commented on test/jdk/javax/swing/ImageIcon/ImageIconTest.java:
```diff
@@ -0,0 +1,121 @@
+            ImageIcon ii = new ImageIcon();
+            ii.setImage((Image)null);
+            throw new RuntimeException("No NPE");
+        } catch (NullPointerException e) {
```
It occurred to me that you could include this in the switch with a new ARGTYPE of "SETIMAGE".
Bit of a misnomer as its the type + the method but then you can also check the non-null invalid data type.

prrace commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -61,6 +61,11 @@
  * from Images. Images that are created from a URL, filename or byte array
  * are preloaded using MediaTracker to monitor the loaded state
  * of the image.
+ * If the image source parameter to a constructor or method is non-null,
```
FWIW when I first suggested this statement I said it needed to cover the setter too
'We probably should add the similar text on setImage().'
I guess we can keep it here now it's here.

prrace commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -61,6 +61,11 @@
  * of the image.
+ * If the image source parameter to a constructor or method is non-null,
+ * but does not reference valid accessible image data,
+ * no exceptions will be thrown but the image will be 'effectively' null,
```
Joe asked for if there's an alternative to 'effectively' null. I'm open to suggestions. "unset" ?

aivanov-jdk replied:
_“unset”_ is also not ideal, it's not clear what it means, especially for constructors where the image isn't set yet.

If I get it right, it refers to the fact that the image in `ImageIcon` instance is non-null, but nothing gets rendered because there's no valid image.

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -61,6 +61,11 @@
  * from Images. Images that are created from a URL, filename or byte array
  * are preloaded using MediaTracker to monitor the loaded state
  * of the image.
+ * If the image source parameter to a constructor or method is non-null,
```
```suggestion
 * of the image.
 *
 * <p>
 * If the image source parameter to a constructor or method is non-null,
```
I suggest putting the discussion on non-null parameters in its own paragraph for ease of reading the javadoc.

prsadhuk replied:
ok

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -61,6 +61,11 @@
+ * but does not reference valid accessible image data,
+ * no exceptions will be thrown but the image will be unset,
+ * as it will have no dimensions and never be drawn, and
+ * {@code getImageLoadStatus()} will report {@see java.awt.MediaTracker#ERRORED}.
```
```suggestion
 * {@code getImageLoadStatus()} will report {@link java.awt.MediaTracker#ERRORED}.
```
You shouldn't use `@see` here, it should either be `@code` or `@link`. I think both references should use the same tag: either `@code` or `@link`; the latter allows quickly navigating to the *method* or constant.

Perhaps, `@link` for the `getImageLoadStatus` method would work better: it allows viewing the description of the method and its purpose; in its turn, `getImageLoadStatus` references possible return values including `MediaTracker.ERRORED` with `@see` tag.

aivanov-jdk replied:
Indeed, this is not supported, the generated javadoc looks like this:

“…and `getImageLoadStatus()` will report .”

Therefore I suggest the following:

```java
 * {@link #getImageLoadStatus()} will report {@code MediaTracker.ERRORED}.
```

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -178,6 +183,7 @@ public ImageIcon (String filename) {
      * of the image.
      * @param location the URL for the image
      * @param description a brief textual description of the image
+     * @throws {@code NullPointerException} if (@code null) URL is passed.
```
```suggestion
     * @throws NullPointerException if a {@code null} URL is passed
```
The exception type is automatically rendered in monospaced font with a link to its description.

You meant to use braces instead of parentheses.

Usually, there's no full stop for `@throws`.

Alternatively, specify the parameter name for more clarity:
```suggestion
     * @throws NullPointerException if {@code location} is {@code null}
```

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -197,6 +203,7 @@ public ImageIcon(URL location, String description) {
      * The icon's description is initialized to be
      * a string representation of the URL.
      * @param location the URL for the image
+     * @throws {@code NullPointerException} if (@code null) URL is passed.
```
```suggestion
     * @throws NullPointerException if a {@code null} URL is passed
```

prsadhuk replied:
ok

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -207,6 +214,7 @@ public ImageIcon (URL location) {
      * Creates an ImageIcon from the image.
      * @param image the image
      * @param description a brief textual description of the image
+     * @throws {@code NullPointerException} if (@code null) Image is passed.
```
```suggestion
     * @throws NullPointerException if a {@code null} image is passed
```
Here, image shouldn't be capitalised, you mean it as the literal _“image”_ rather than a type.

I still prefer specifying the parameter name:
```suggestion
     * @throws NullPointerException if {@code image} is {@code null}
```

prsadhuk replied:
ok

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -218,6 +226,7 @@ public ImageIcon(Image image, String description) {
      * If the image has a "comment" property that is a string,
      * then the string is used as the description of this icon.
      * @param image the image
+     * @throws {@code NullPointerException} if (@code null) Image is passed.
```
```suggestion
     * @throws NullPointerException if {@code image} is {@code null}
```

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -241,6 +250,7 @@ public ImageIcon (Image image) {
      * @param  imageData an array of pixels in an image format supported
      *         by the AWT Toolkit, such as GIF, JPEG, or (as of 1.3) PNG
      * @param  description a brief textual description of the image
+     * @throws {@code NullPointerException} if (@code null) imageData is passed.
```
```suggestion
     * @throws NullPointerException if {@code imageData} is {@code null}
```

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -264,6 +274,7 @@ public ImageIcon (byte[] imageData, String description) {
      *
      * @param  imageData an array of pixels in an image format supported by
      *             the AWT Toolkit, such as GIF, JPEG, or (as of 1.3) PNG
+     * @throws {@code NullPointerException} if (@code null) imageData is passed.
```
```suggestion
     * @throws NullPointerException if {@code imageData} is {@code null}
```

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -368,6 +379,7 @@ public Image getImage() {
     /**
      * Sets the image displayed by this icon.
      * @param image the image
+     * @throws {@code NullPointerException} if (@code null) Image is passed.
```
```suggestion
     * @throws NullPointerException if {@code image} is {@code null}
```

aivanov-jdk commented on test/jdk/javax/swing/ImageIcon/ImageIconTest.java:
```diff
@@ -0,0 +1,123 @@
+        try (FileOutputStream fos = new FileOutputStream(imgName)) {
+            fos.write(invalidData);
+        }
+        String fileName = (new File(System.getProperty("test.src", "."), imgName)).getName();
```
Why do we need jumping over the hoops?

You're creating an image, put it into the current directory that defaults to `scratch` directory in jtreg. The `scratch` directory is automatically cleaned by jtreg between tests, thus you won't even leave files in the source tree of JDK.

prsadhuk replied:
used file.deleteOnExit

aivanov-jdk commented on test/jdk/javax/swing/ImageIcon/ImageIconTest.java:
```diff
@@ -0,0 +1,123 @@
+                boolean passed = false;
+                try {
+                    switch (a) {
+                       case FILE :
+                           expected = false;
+                           String s = (v == ArgVal.NULL) ? null : fileName;
+                           new ImageIcon(s);
+                           passed = true; // no exception expected for this case
+                           break;
```
Does it make sense to create richer enum values which combine the expected state?

aivanov-jdk commented on test/jdk/javax/swing/ImageIcon/ImageIconTest.java:
```diff
@@ -0,0 +1,123 @@
+import java.awt.Toolkit;
+import javax.swing.ImageIcon;
+
+public class ImageIconTest {
```
```suggestion
public class ImageIconNullParametersTest {
```
Be more specific?

prsadhuk replied:
it is not just testing for NullParameters

aivanov-jdk reviewed (changes requested):
At what point did we decide not to modify `Image.setImage` to accept `null` as the parameter? It made perfect sense.

I wonder if it's possible to use JUnit a test framework in jtreg?

aivanov-jdk commented:
Currently, [JDK-8159055](https://bugs.openjdk.org/browse/JDK-8159055) is an enhancement which made sense while the goal of this PR was to change `Image.setImage` so that it accepts `null` as parameter.

After a month of discussions, that solution has been abandoned, instead it is stated that `Image.setImage` will throw `NullPointerException` if `image` parameter is `null`.

I think the type of JDK-8159055 should be changed to *bug*.

aivanov-jdk reviewed (approved):
`ImageIcon` changes look good to me.

I admit I can't understand what the test is doing, and I don't have to time to study the test thoroughly.

However, I'm still for documenting null and invalid parameters under a separate bug id; under this bug id modify `Image.setImage` to accept `null` as it's reported in the bug ([JDK-8159055](https://bugs.openjdk.org/browse/JDK-8159055)) and as it was done in this PR initially.

prsadhuk commented on test/jdk/javax/swing/ImageIcon/ImageIconTest.java:
```diff
@@ -0,0 +1,125 @@
+        }
+        File file = new File(imgName);
+        file.deleteOnExit();
+        String fileName = file.getName();
```
yes removed

prrace commented:
> Currently, [JDK-8159055](https://bugs.openjdk.org/browse/JDK-8159055) is an enhancement which made sense while the goal of this PR was to change `Image.setImage` so that it accepts `null` as parameter.
> 
> After a month of discussions, that solution has been abandoned, instead it is stated that `Image.setImage` will throw `NullPointerException` if `image` parameter is `null`.
> 
> I think the type of JDK-8159055 should be changed to _bug_.

I agree. Not saying how null is handled is a bug. So fixing it is a bug fix, not an enhancement.

prrace commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -63,6 +63,13 @@
  * <p>
+ * If the image source parameter to a constructor or method is non-null,
+ * but does not reference valid accessible image data,
+ * no exceptions will be thrown but the image will be unset,
```
"will behave as if unset" ?

prsadhuk replied:
@prrace Any objection to my suggestion of statement? I guess that is succinct and describe what's the end result!!
`no exceptions will be thrown but the image will behave as if unset/uninitialized/null` doesn't sound right in my opinion as then user need to know "how it would behave" when image is unset or null which is basically image icon not getting rendered so I thought to cut the chase and describe the final result

prrace replied:
> “will behave as if the image is null”

That's got the same mention of null that Joe was questioning

>"no exceptions will be thrown but no image icon will be rendered"

That seems OK to me. I'd just drop the word 'icon'.

prsadhuk replied:
Updated the class-level javadoc

prrace commented on src/java.desktop/share/classes/javax/swing/ImageIcon.java:
```diff
@@ -63,6 +63,14 @@
  * of the image.
  *
  * <p>
+ * If the image source parameter to a constructor or method is non-null,
+ * but does not reference valid accessible image data,
+ * no exceptions will be thrown but no image will be rendered
+ * even though {@link #getImage()} will return a non-null value,
+ * as the image will have no dimensions
+ * and {@link #getImageLoadStatus()} will report {@code MediaTracker.ERRORED}.
```
I prefer one sentence because these aren't disjoint statements.

aivanov-jdk replied:
Okay. They're still the same statements in a few sentences rather than in one long sentence.

aivanov-jdk commented on test/jdk/javax/swing/ImageIcon/ImageIconTest.java:
```diff
@@ -0,0 +1,150 @@
+                boolean expected = true;
+                boolean passed = false;
+                try {
+                    switch (a) {
+
+                       case FILE :
+
+                           expected = false;
+                           String s = (v == ArgVal.NULL) ? null : imgName;
+                           new ImageIcon(s);
+                           passed = true; // no exception expected for this case
+                           break;
+
+                       case URL :
```
Was adding so many blank lines intentional?

aivanov-jdk reviewed (approved):
Looks good… except for lots of unexpected blank lines in the test code.

prrace commented on test/jdk/javax/swing/ImageIcon/ImageIconTest.java:
```diff
@@ -0,0 +1,125 @@
+                           break;
+                       case IMAGE :
+                           if (v == ArgVal.NULL) {
+                               new ImageIcon((Image)null);
```
Personally, I do not like having a space there.

prsadhuk commented:
/integrate
